firedrake/interpolation.py

```diff
@@ -33,12 +33,50 @@
 __all__ = (
     "interpolate",
     "Interpolator",
+    "Interpolate",
     "DofNotDefinedError",
     "CrossMeshInterpolator",
     "SameMeshInterpolator",
 )
 
 
+class Interpolate(ufl.Interpolate):
+
+    def __init__(self, expr, v, interp_data=None):
+        r""" Symbolic representation of the interpolation operator.
+
+        :arg expr: a UFL expression to interpolate.
+        :arg v: the :class:`.FunctionSpace` to interpolate into or the :class:`.Coargument`
+                defined on the dual of the :class:`.FunctionSpace` to interpolate into.
+        """
+
+        # Check function space
+        if isinstance(v, functionspaceimpl.WithGeometry):
+            v = Argument(v.dual(), 0)
+            v2 = extract_arguments(expr)
+            if v2 and v2[0].number() == 0:
+                # Cope with the different convention of Interp and Inteprolator:
+                #  -> Interpolate(Argument(V1, 1), Argument(V2.dual(), 0))
+                #  -> Interpolator(Argument(V1, 0), V2)
+                expr = replace(expr, {v2[0]: Argument(v2[0].function_space(),
+                                                      number=1,
+                                                      part=v2[0].part())})
+
+        # Get the primal space (V** = V)
+        vv = v if not isinstance(v, ufl.Form) else v.arguments()[0]
+        self._function_space = vv.function_space().dual()
+        ufl.Interpolate.__init__(self, expr, v)
+
+        # Interpolate data (e.g. subset or access)
+        self.interp_data = interp_data if interp_data else {}
+
+    def function_space(self):
+        return self._function_space
+
+    def _ufl_expr_reconstruct_(self, expr, v=None, interp_data=None):
+        return ufl.Interpolate._ufl_expr_reconstruct_(self, expr, v=v, interp_data=interp_data or self.interp_data)
```

Comment:
We override this because `interp_data` is a Firedrake thing. As for the other arguments (`expr` and `v`), they are handled by the parent class, which will do that.

Reply:
I see. But if `interp_data` is a Firedrake thing then why are we passing it to the UFL `Interpolate` constructor. Won't that error?

Reply:
It accepts extra optional arguments (see https://github.com/firedrakeproject/ufl/blob/master/ufl/core/interpolate.py#L61)

Reply:
I see. Thanks!